messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java

```diff
@@ -83,15 +94,20 @@
  * </pre>
  *
  * @author Rob Winch
+ * @author DingHao
  * @since 4.0
  */
 public final class AuthenticationPrincipalArgumentResolver implements HandlerMethodArgumentResolver {
 
 	private SecurityContextHolderStrategy securityContextHolderStrategy = SecurityContextHolder
 		.getContextHolderStrategy();
 
+	private final Map<MethodParameter, Annotation> cachedAttributes = new ConcurrentHashMap<>();
+
 	private ExpressionParser parser = new SpelExpressionParser();
 
+	private PrePostTemplateDefaults defaults = new PrePostTemplateDefaults();
```

Comment:
I think we'll want to introduce a new class, maybe `AuthenticationPrincipalTemplateDefaults`. By the name and JavaDoc `PrePostTemplateDefaults` is intended for the pre/post annotations.

Reply:
Done